pkg/agent/billing.go

```diff
@@ -0,0 +1,181 @@
+	pushTicker := time.NewTicker(time.Second * time.Duration(conf.PushEverySeconds))
+	defer pushTicker.Stop()
+
+	state := billingMetricsState{
+		historical: make(map[billingMetricsKey]vmMetricsHistory),
+		present:    make(map[billingMetricsKey]vmMetrics),
+		lastTime:   nil,
+	}
+
+	state.collect(conf, store)
+
+	for {
+		select {
+		case <-collectTicker.C:
+			state.collect(conf, store)
+		case <-pushTicker.C:
+			batch := state.drainIntoBatch(&client)
+			if err := pushBillingEvents(conf, batch); err != nil {
+				return fmt.Errorf("Error pushing billing events: %w", err)
+			}
+		case <-backgroundCtx.Done():
+			// If we're being shut down, push the latests events we have before returning.
+			batch := state.drainIntoBatch(&client)
+			if err := pushBillingEvents(conf, batch); err != nil {
+				return fmt.Errorf("Error pushing billing events: %w", err)
+			}
+			return nil
+		}
+	}
+}
+
+func (s *billingMetricsState) collect(conf *BillingConfig, store *util.WatchStore[vmapi.VirtualMachine]) {
+	now := time.Now()
+
+	old := s.present
+	s.present = make(map[billingMetricsKey]vmMetrics)
+	for _, vm := range store.Items() {
+		endpointID, ok := vm.Labels[EndpointLabel]
+		if !ok {
+			continue
+		}
+
+		key := billingMetricsKey{
+			uid:        vm.UID,
+			endpointID: endpointID,
+		}
+		presentMetrics := vmMetrics{
+			cpu: uint16(*vm.Spec.Guest.CPUs.Use),
+		}
+		if oldMetrics, ok := old[key]; ok {
+			// The VM was present from s.lastTime to now. Add a time slice to its metrics history.
+			timeSlice := metricsTimeSlice{
+				metrics: vmMetrics{
+					// strategically under-bill by assigning the minimum to the entire time slice.
+					cpu: util.Min(oldMetrics.cpu, presentMetrics.cpu),
+				},
+				// note: we know s.lastTime != nil because otherwise old would be empty.
+				startTime: *s.lastTime,
+				endTime:   now,
+			}
+
+			vmHistory, ok := s.historical[key]
+			if !ok {
+				vmHistory = vmMetricsHistory{slices: nil}
+			}
+			vmHistory.slices = append(vmHistory.slices, timeSlice)
+			s.historical[key] = vmHistory
+		}
+
+		s.present[key] = presentMetrics
+	}
+
+	s.lastTime = &now
+}
+
+func (s *billingMetricsState) drainIntoBatch(client *billing.Client) *billing.Batch {
+	batch := client.NewBatch()
+
+	for key, history := range s.historical {
+		history.forEachMergedSlice(func(slice metricsTimeSlice) {
+			batch.AddIncrementalEvent(billing.IncrementalEvent{
+				IdempotencyKey: uuid.NewString(),
+				MetricName:     MetricNameCPU,
+				Type:           "", // set on JSON marshal
+				EndpointID:     key.endpointID,
+				StartTime:      slice.startTime,
+				StopTime:       slice.endTime,
+				Value:          int(slice.metrics.cpu),
+			})
+		})
+	}
+
+	s.historical = make(map[billingMetricsKey]vmMetricsHistory)
+	return batch
+}
+
+func (h vmMetricsHistory) forEachMergedSlice(process func(metricsTimeSlice)) {
+	if len(h.slices) == 0 {
+		return
+	}
+
+	last := h.slices[0]
+	for _, slice := range h.slices[1:] {
```

Comment:
range is kind of cute behind the scenes, so I think `slice` is always the same pointer object they just move what it points to? so I often (perhaps with an abundance of caution) use the index rather than the value output of range, a la: 

```go
for idx := range h.slices {
    cur := h.slices[idx] 
}
```

Though admittedly you're also doing someithing clever by starting at the second element, so maybe c-style rather than range is the way to go.

Reply:
in reverse order:

> admittedly you're also doing someithing clever by starting at the second element, so maybe c-style rather than range is the way to go.

I'll think about it for a bit. I like the readability of range syntax, but I see where you're coming from.

> I think `slice` is always the same pointer object they just move what it points to?

_kind of_ — IIRC it's more accurate to desugar this range as:

```go
{
    var slice metricsTimeSlice
    for i := 1; i < len(slice); i += 1 {
        slice = h.slices[i]
        // -- insert for loop body here --
    }
}
```

so you can run into issues when you create a reference to either `i` or `slice` that lives longer than the loop body, either explicitly with `&slice` or implicitly with a pointer method or anonymous function that captures it.